doc/source/development/rfc/rfc99_geometry_coordinate_precision.rst

```diff
@@ -0,0 +1,627 @@
+    }
+
+
+Corresponding additions at the C API level:
+
+.. code-block:: c
+
+    /** Value for a unknown coordinate precision. */
+    #define OGR_GEOM_COORD_PRECISION_UNKNOWN 0
+
+    /** Opaque type for OGRGeomCoordinatePrecision */
+    typedef struct OGRGeomCoordinatePrecision *OGRGeomCoordinatePrecisionH;
+
+    OGRGeomCoordinatePrecisionH CPL_DLL OGRGeomCoordinatePrecisionCreate(void);
+    void CPL_DLL OGRGeomCoordinatePrecisionDestroy(OGRGeomCoordinatePrecisionH);
+    double CPL_DLL
+        OGRGeomCoordinatePrecisionGetXYResolution(OGRGeomCoordinatePrecisionH);
+    double CPL_DLL
+        OGRGeomCoordinatePrecisionGetZResolution(OGRGeomCoordinatePrecisionH);
+    double CPL_DLL
+        OGRGeomCoordinatePrecisionGetMResolution(OGRGeomCoordinatePrecisionH);
+    char CPL_DLL **
+        OGRGeomCoordinatePrecisionGetFormats(OGRGeomCoordinatePrecisionH);
+    CSLConstList CPL_DLL OGRGeomCoordinatePrecisionGetFormatSpecificOptions(
+        OGRGeomCoordinatePrecisionH, const char *pszFormatName);
+    void CPL_DLL OGRGeomCoordinatePrecisionSet(OGRGeomCoordinatePrecisionH,
+                                               double dfXYResolution,
+                                               double dfZResolution,
+                                               double dfMResolution);
+    void CPL_DLL OGRGeomCoordinatePrecisionSetFromMetre(OGRGeomCoordinatePrecisionH,
+                                                        OGRSpatialReferenceH hSRS,
+                                                        double dfXYMetreResolution,
+                                                        double dfZMetreResolution,
+                                                        double dfMResolution);
+    void CPL_DLL OGRGeomCoordinatePrecisionSetFormatSpecificOptions(
+        OGRGeomCoordinatePrecisionH, const char *pszFormatName,
+        CSLConstList papszOptions);
+
+OGRGeomFieldDefn class
+++++++++++++++++++++++
+
+The existing :cpp:class:`OGRGeomFieldDefn` is extended with a new
+OGRGeomCoordinatePrecision member, and associated getter and setter methods.
+
+.. code-block:: c++
+
+    class OGRGeomFieldDefn
+    {
+        public:
+            const OGRGeomCoordinatePrecision& GetCoordinatePrecision() const;
+
+            void SetCoordinatePrecision(const OGRGeomCoordinatePrecision& prec);
+
+        private:
+            OGRGeomCoordinatePrecision m_oCoordPrecision{};
+    };
+
+New corresponding C API:
+
+.. code-block:: c
+
+    OGRGeomCoordinatePrecisionH
+        CPL_DLL OGR_GFld_GetCoordinatePrecision(OGRGeomFieldDefnH);
+    void CPL_DLL OGR_GFld_SetCoordinatePrecision(OGRGeomFieldDefnH,
+                                                 OGRGeomCoordinatePrecisionH);
+
+WKB export
+++++++++++
+
+WKB export methods will be modified in a similar way as in the prototype
+https://github.com/OSGeo/gdal/pull/6974 to nullify least significant bits from
+the precision specifications.
```

Comment:
I wonder, if I understand correctly (which might not be the case!), the coordinates will be rounded as such... 
However, simply rounding coordinates can/will lead to invalid polygons (e.g. self-intersections, polygon rings touching,...).
So, I wonder, why not use GEOS set_precision to round the coordinates in the export methods?

Reply:
> the coordinates will be rounded as such...

yes

> However, simply rounding coordinates can/will lead to invalid polygons (e.g. self-intersections, polygon rings touching,...).

This may indeed be possible if users indicate a precision not consistent of the actual point spacing of their geometries

> So, I wonder, why not use GEOS set_precision to round the coordinates in the export methods?

Good suggestion. However there would be a cost in doing so systematically. But I'm inclined to add that a GEOSGeom_setPrecision_r() step in a ogr2ogr workflow when the user explicitly specifies -xyRes, which can indeed causes generation of invalid geometries.  For other workflows, the assumption is that the coordinate precision of the source is correct.

Reply:
RFC text and implementation updated with the above

Reply:
> Maybe a configuration option could be introduced to give the user the option to apply set_precision systematically when writing?

It is not obvious to me that this is needed at that point, because either people use ogr2ogr and it takes care of that, or they use individual steps of the OGR API (CreateLayer(), CreateFeature()) and can explicitly invoke OGRGeometry::SetPrecision() if needed. A config option to do that systematically (not clear at that point if that would be in each driver that honours coordinate precision, or in in the generic CreateFeature() code) could potentially be added as a further enhancement

Reply:
I was thinking about the drivers that honour the coordinate precision: if the precision is set and the configuration option is enabled, the validity of the geometries are guaranteed by gdal but a performance penalty can occur. If the the precision is set and configuration option is not enabled... it is the user's responsibility to make sure the geometries are rounded in a way they are valid as otherwise they can end up becoming invalid when saved.

Reply:
oh well, I might consider doing this, but for GeoPackage (or formats that would use the same strategy), the value that is written will not be "exactly" the one after rounding to the precision, due to zeroing out least significant bits. It will be *compatible* with the requested precision however. That is if you ask for a precision of let's say 0.1 and you encode 0.1, the value you get when reading back will be 0.125 for example. Would that be a problem? My intuition would be that if the geometries are valid with respect to the 0.1 precision, those small changes of coordinate values < 0.1 shouldn't induce new in-validities, but I might be wrong...
If so, I was thinking about potentially also applying the precision when reading back so that 0.125 gets seen as 0.1, but that cost might be undesirable, and would only be done by the GDAL reader, so likely not a good option.
Or perhaps we should apply the zeroing out of LSBs in GeoPackage WKB only if a specific creation option is specified like DISCARD_COORD_LSB=YES ? (not sure which default should be used)

Reply:
Good questions. I follow the same intuition regarding the probable limited impact of zeroing LSBs, but I'm not at all a specialist in floating point representations, so not sure at all.

Applying precision when reading indeed sounds as not ideal... hopefully it can be avoided.

Maybe it is worth a test to see if zeroing out the LSB's after using GEOS set_precision is still having a significant impact?

Reply:
> Maybe it is worth a test to see if zeroing out the LSB's after using GEOS set_precision is still having a significant impact?

on a quick check, as expected, the zeroing out of the LSB's can be reversed to the initial "decimal" precision, since we are conservative on the number of zeroed LSB's, and the geometry is still valid. However I've revered the default behaviour of the GPKG driver to not do that zeroing out by default. See the last commit (4fb623301dc3f909a9100a4a7041bfbd8a255a2e) of this PR for the updates, as well as the new OGR_APPLY_GEOM_SET_PRECISION configuration option that is taken into account by the generic code path of CreateFeature() to run OGRGeometry::SetPrecision().

Reply:
Looks great.